### PR TITLE
Add CodeQL support

### PR DESCRIPTION
**Pull request by makubacki**
## Description

Adds a CodeQL plugin that supports CodeQL in the build system.

1. CodeQlBuildPlugin - Generates a CodeQL database for a given build.
2. CodeQlAnalyzePlugin - Analyzes a CodeQL database and interprets results.
3. External dependencies - Assist with downloading the CodeQL CLI and making it available to the CodeQL plugins.
4. MuCodeQlQueries.qls - A Project Mu CodeQL query set run against Project Mu code.
5. Readme.md - A comprehensive readme file to help:
   - Platform integrators understand how to configure the plugin
   - Developers understand how to modify the plugin
   - Users understand how to use the plugin

Read Readme.md for additional details.

- [ ] Breaking change?
  - Will this change break pre-existing builds or functionality without action being taken?
  **No**

## How This Was Tested

1. Verified CodeQlBuildPlugin generates a CodeQL database against every mu_basecore package
2. Verified all configuration options in CodeQlBuildPlugin work
3. Verified CodeQlAnalyzePlugin produces a SARIF result file as expected
4. Verified all configuration options in CodeQlAnalyzePlugin work
5. Verified SARIF results are read correctly
6. Verified all external Web dependencies work

## Integration Instructions

Read the Readme.md file added in this change.

Note: The plugin will be used to enable CodeQL in this repo in a separate, future
change after some additional CodeQL fixes have gone into the repo.

Signed-off-by: Michael Kubacki <michael.kubacki@microsoft.com>

### CODE REVIEW

File: .pytool/Plugin/CodeQL/Readme.md [hunk 1]
```diff
@@ -0,0 +1,288 @@
+# CodeQL Plugin
+
+The set of CodeQL plugins provided, include two main plugins:
```
Comment (os-d) [inline .pytool/Plugin/CodeQL/Readme.md]: nit: no comma

**Review by os-d — APPROVED** [review]

File: .pytool/Plugin/CodeQL/Readme.md (hunk 1 above)
Comment (makubacki) [reply .pytool/Plugin/CodeQL/Readme.md]: Removed in commit https://github.com/microsoft/mu_basecore/pull/163/commits/e1b7977e9cdfbc6aa7bcc37a7593be548b8c6d48

File: .pytool/Plugin/CodeQL/Readme.md [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+|     934 MB     | 415 MB |  290 MB |
+
+Therefore, the following is recommended:
+
+1. **Ideal** - Create container images for build agents and install the CodeQL CLI for the container OS into the
+   container.
+2. Leverage host-OS detection (e.g. [`GetHostInfo()`](https://github.com/tianocore/edk2-pytool-library/blob/42ad6561af73ba34564f1577f64f7dbaf1d0a5a2/edk2toollib/utility_functions.py#L112))
+to set the scope for the appropriate operating system. This will download the much smaller OS-specific application.
+
+> _NOTE:_ You should never have more than one CodeQL external dependency scope enabled at a time.
```
Comment (apop5) [inline .pytool/Plugin/CodeQL/Readme.md:155]: do the scripts generate an error if more than one CodeQL external dependency is scoped?

**Review by apop5 — APPROVED** [review]

File: .pytool/Plugin/CodeQL/Readme.md (hunk 2 above)
Comment (makubacki) [reply .pytool/Plugin/CodeQL/Readme.md:155]: No. I'm not sure plugins have direct access to scopes. But each ext dep ultimately sets the same environment variable so I wouldn't anticipate any harm other than wasted time.

**Review by cfernald — APPROVED** [review]

**Comment by makubacki** [comment]
I'm waiting to submit this PR until a few PRs in edk2-pytool-library and edk2-pytool-extensions complete. Once that happens, I will update pip-requirements.txt in this PR with the release versions and complete this PR.

**Pull request merged**